File: Trabajo_9/PY/Ejercicios/Ejercicio_3.py
from abc import ABC, abstractmethod

class Empleado(ABC):
    @abstractmethod
    def calcularSalario(self):
        pass
    def agregarBono(self):
        pass

class Asalariados(Empleado):
    def __init__(self, nombre, apellido, codigo):
        self.nombre = nombre
        self.apellido = apellido
        self.codigo = codigo
        self.bono = 0
        self.salario = 1000

    def calcularSalario(self):
        print ("El salario del empleado es: ", self.salario+self.bono)

    def agregarBono(self):
        menuBono()
        op = 0
        while True:
            op = int(input("Ingrese una Opcion 1-4: "))
            if (op >0 and op <= 4):
                break
            else:
                print("Opcion invalida")
        if op == 1:
            self.bono = 0
        elif op == 2:
            self.bono = 50
        elif op == 3:
            self.bono = 100
        elif op == 4:
            self.bono = 200

    def __str__(self):
        return (f"Empleado {self.nombre} {self.apellido}\n"
                f"  Codigo {self.codigo}\n"
                f"  Salario s/{self.salario}\n"
                f"  Bono {self.bono}")

class PorHora(Empleado):
    def __init__(self, nombre, apellido, codigo, horas):
        self.nombre = nombre
        self.apellido = apellido
        self.codigo = codigo
        self.horas = horas
        self.bono = 0


    def calcularSalario(self):
        print(f"Salario: {self.horas*50}")

    def agregarBono(self):
        self.bono = self.horas*8

    def __str__(self):
        return (f"Empleado {self.nombre} {self.apellido}\n"
                f"  Codigo: {self.codigo}\n"
                f"  Salario: {self.horas*50}\n"
                f"  Bono: {self.bono}")

class Comisionista(Empleado):
    def __init__(self, nombre, apellido, codigo, nVentas):
        self.nombre = nombre
        self.apellido = apellido
        self.codigo = codigo
        self.nVentas = nVentas
        self.bono = 0

    def calcularSalario(self):
        return self.nVentas*50

    def agregarBono(self):
        self.bono = self.nVentas*8

    def __str__(self):
        print(f"Comisionista {self.nombre} {self.apellido}\n"
              f"  Codigo: {self.codigo}\n"
              f"  Ventas: {self.nVentas}\n"
              f"  Bono: {self.bono}\n"
              f"  Salario: {self.calcularSalario()}")


def menuBono():
    print("BONOS \n"
          " 1 -> s/0 \n"
          " 2 -> s/50 \n"
          " 3 -> s/100 \n"
          " 4 -> s/200 ")

def menuEmpleados():
    print("Empleados \n"
          "1 ->Asalariado\n"
          "2 -> PorHora\n"
          "3 -> Comisionista\n"
          "4 -> Salir")

if __name__ == "__main__":
    while True:
        menuEmpleados()
        op = int(input("Ingrese una opcion 1-4: "))
        if (op <=0 or op > 4):
            print("Opcion invalida")
            continue
        elif (op == 1):
            asalar = Asalariados("Juan","Martinez",341)
            asalar.agregarBono()
            print(asalar)
        elif (op == 2):
            por_hora = PorHora("Juan","Martinez",211,10)
            por_hora.agregarBono()
            print(por_hora)
        elif (op == 3):
            comisio = Comisionista("Juan","Martinez",341,200)
            comisio.agregarBono()
            print(comisio)
        else:
            print("Saliendo de Programa")
            break

File: Trabajo_9/CPP/Ejercicios/Ejercicio_3.cpp
#include <iostream>
using namespace std;

//Funciones extra
void menu_bono(){
    cout << "Ingrese el bono correspondiente: " << endl;
    cout << "0 -> s/0" << endl;
    cout << "1 -> s/50" << endl;
    cout << "2 -> s/100" << endl;
    cout << "3 -> s/150" << endl;
};

bool validacion(int i, int end){
    for (int x = 0; x < end; x++){
        if ( x == i ){
            return true;
        }
    }
    return false;
}

//Clases
class Empleados{
public:
    virtual void calcularSalario() const = 0;
    virtual void agregarBono() = 0;
};

class Asalariados : public Empleados {
private:
    string nombre;
    string apellido;
    int codigo;
    int bono;
    int const salario = 1000;

public:
    Asalariados(string n, string l, int code, int b = 0) {
        this->nombre = n;
        this->apellido = l;
        this->codigo = code;
        this->bono = b;
    }

    void calcularSalario() const override {
        cout << "Salario c/bono -> " << salario + bono << endl;
    }

    void agregarBono() override {
        menu_bono();
        int extra;
        cin >> extra;
        while (!validacion(extra, 4)) {
            cout << "Ingrese una opcion valida: " << endl;
            cin >> extra;
        }

        switch (extra) {
            case 0:
                break;
            case 1:
                this->bono = 50;
                break;
            case 2:
                this->bono = 100;
                break;
            case 3:
                this->bono = 150;
                break;
        }
    }

    friend ostream& operator << (ostream& ou, Asalariados asalario){
        ou << "Datos de empleado de nombre " << asalario.nombre + asalario.apellido << " :" << endl;
        ou << "Codigo        -> " << asalario.codigo << endl;
        ou << "Sueldo base   -> " << asalario.salario << endl;
        asalario.agregarBono();
        asalario.calcularSalario();
        return ou;
    }
};

class Por_hora: public Empleados {
private:
    string nombre;
    string apellido;
    int codigo;
    int horas;
    int bono;

public:
    Por_hora(string n, string l, int code, int hours = 0, int b = 0) {
        this->nombre = n;
        this->apellido = l;
        this->codigo = code;
        this->horas = hours;
        this->bono = b;
    }

    void calcularSalario() const override {
        cout << "Salario -> " << (horas * 50) + bono << endl;
    }

    void agregarBono() override {
        int extra = horas * 25;
        this -> bono = extra;
        cout << "Siendo las horas trabajadas -> " << horas << " su bono es de " << extra << endl;
    }

    friend ostream& operator << (ostream& ou, Por_hora porHora){
        ou << "Datos de empleado de nombre " << porHora.nombre + porHora.apellido << " :" << endl;
        ou << "Codigo           -> " << porHora.codigo << endl;
        ou << "Horas Trabajadas -> " << porHora.horas << endl;
        porHora.agregarBono();
        porHora.calcularSalario();
        return ou;
    }
};


class Comisionista: public Empleados{
private:
    string nombre;
    string apellido;
    int codigo;
    int const precio_maquina = 1000;
    int maquinas_vendidas;
    int bono;

public:
    Comisionista(string n, string l, int code, int machines = 0, int b = 0) {
        this->nombre = n;
        this->apellido = l;
        this->codigo = code;
        this->maquinas_vendidas= machines;
        this->bono = b;
    }

    void calcularSalario() const override {
        cout << "Salario -> " << maquinas_vendidas*(precio_maquina * 0.25) + bono << endl;
    }

    void agregarBono() override {
        int extra = maquinas_vendidas * 100;
        this -> bono = extra;
        cout << "Siendo las maquinas vendidas -> " << maquinas_vendidas << " su bono es de " << extra << endl;
    }

    friend ostream& operator << (ostream& ou, Comisionista comisionista){
        ou << "Datos de empleado de nombre " << comisionista.nombre + comisionista.apellido << " :" << endl;
        ou << "Codigo            -> " << comisionista.codigo << endl;
        ou << "Maquinas Vendidas -> " << comisionista.maquinas_vendidas << endl;
        comisionista.agregarBono();
        comisionista.calcularSalario();
        return ou;
    }
};

void menu_empleados() {
    cout << "Crear: " << endl;
    cout << "0 -> Finalizar Programa" << endl;
    cout << "1 -> Asalariado" << endl;
    cout << "2 -> Por hora" << endl;
    cout << "3 -> Comisionista" << endl;
}

int main() {
    while (true) {
        menu_empleados();
        int opc;
        cout << "Ingrese una opcion: " << endl;
        cin >> opc;
        while (!validacion(opc, 4)) {
            cout << "Ingrese una opcion valida: " << endl;
            cin >> opc;
        }
        switch (opc) {
            case 0: {
                return 0;
            }
            case 1: {
                Asalariados asalariado("Grossman ", "Vargas", 1);
                cout << asalariado;
                break;
            }
            case 2: {
                Por_hora porHora("Yhosfer ", "Quispe", 2, 40);
                cout << porHora;
                break;
            }
            case 3: {
                Comisionista comisionista("Luis ", "Pozu", 3, 10);
                cout << comisionista;
                break;
            }
        }
    }
}